Feature/extend dataset tests

nshumoogum opened the pull request:
### What

Extend test coverage for almost all endpoints on the dataset API and refactor code to use gherkin (given, when, then). Test coverage for version and instance dimensions not extended due to restructuring the endpoints, which will mean a revamp of the spec for `/datasets/...../dimensions`, `/instances/....../dimensions`, `/datasets/...../options` and `/instances/....../options`.

### How to review

Make sure you have checkout `feature/fix-dataset-state-changes` on dp-dataset-api, run the tests by doing the following:

1) Run dataset API
2) cd `<path>/dp-api-tests/datasetAPI`
3) Run `go test ./...`

Make sure tests make sense and follow spec

### Who can review

Team B or Manoj

nshumoogum committed:
Extend dataset api tests to cover more scenarios

All tests cases for dataset API now involve calling mongo to check
created or updated resources instead of relying on the response to a
request. This not only makes it clear that fields have been
updated/created in our datastore but also due to manipulation by the API
depending on user and request the document may appear different in
datastore; also updates can result in other resources being updated, so
these need to be checked to see whether they have been updated
correctly.

nshumoogum committed:
Merge branch 'cmd-develop' into feature/extend-dataset-tests

nshumoogum committed:
Remove commented out code

Matt-Guest commented on datasetAPI/get_dataset_dimensions_test.go:
```diff
@@ -216,7 +109,126 @@ func TestGetDimensions_Failed(t *testing.T) {
 
-func checkDimensionsResponse(response *httpexpect.Object) {
+// TODO Reinstate these tests when code has been fixed
+// func TestGetDimensions_Failed(t *testing.T) {
```
Could you just add `t.Skip("Skipping tests due to.....")` instead of comment out the tests. Is it possible to add a comment on why is the test failing.

nshumoogum replied:
👍

Matt-Guest commented on datasetAPI/get_dimension_options_test.go:
```diff
@@ -246,7 +131,134 @@ func TestGetDimensionOptions_Failed(t *testing.T) {
 
-func checkTimeDimensionResponse(response *httpexpect.Object) {
+// TODO Reinstate these tests when code has been fixed
+// func TestGetDimensionOptions_Failed(t *testing.T) {
```
Same again here

Matt-Guest commented on datasetAPI/get_editions_test.go:
```diff
@@ -47,94 +52,98 @@ func TestSuccessfullyGetListOfDatasetEditions(t *testing.T) {
 
-	if err := mongo.TeardownMany(d); err != nil {
-		if err != mgo.ErrNotFound {
+	Convey("Given a dataset has and edition that is published and one that is unpublished", t, func() {
```
typo "Given a dataset has and edition" > "Given a dataset has an edition"

Matt-Guest commented on datasetAPI/get_instance_dimensions_test.go:
```diff
@@ -104,53 +111,49 @@ func TestGetInstanceDimensions_ReturnsAllDimensionsFromAnInstance(t *testing.T)
-		}
-	}
-}
+// TODO Reinstate tests once endpoint is fixed
```
Here as well

Matt-Guest commented on datasetAPI/get_versions_test.go:
```diff
@@ -200,11 +240,59 @@ func checkVersionResponse(response *httpexpect.Object) {
+		Update:     validAssociatedInstanceData(datasetID, edition, unpublishedInstanceID),
+	}
+
+	docs = append(docs, datasetDoc, editionDoc, instanceDoc, unpublishedInstanceDoc) //, instanceDoc)
```
commented out code

Matt-Guest commented on datasetAPI/initialise.go:
```diff
@@ -11,14 +11,14 @@ import (
+	// edition                = "2017"
+	// instanceID             = "789" // This maybe known as the version Id
+	// dimensionID            = "999"
+	// dimensionOptionID      = "888"
```
commented out code

nshumoogum committed:
Update response error message against dataset API endpoints

Matt-Guest commented on datasetAPI/put_dataset_test.go:
```diff
@@ -2,88 +2,264 @@ package datasetAPI
 	})
 }
+
+// ------------------------------------------------------------------------
```
Is this needed?

Matt-Guest commented on datasetAPI/put_version_test.go:
```diff
@@ -10,78 +11,650 @@ import (
+			os.Exit(1)
+		}
+
+		// TODO reinstate test once dataset API has been fixed
```
Can we just skip this test?

Matt-Guest commented:
All tests are passing on my local box 👍

nshumoogum committed:
Uncomment out tests, instead skip these tests

Tests being skipped due to functionality missing from dataset API and
this could include refactoring those endpoints.
